sway-fmt-v2/src/utils/comments.rs

```diff
@@ -69,6 +100,308 @@ fn collect_comments_from_token_stream(
+}
+
+// A CommentWithContext is the Comment and the offset before it. The offset can be between the (from) item we searched for this comment or from the last comment inside range
+type CommentWithContext = (Comment, String);
+
+/// Returns a list of comments between given spans. For each comment returns the Context
+/// Context of a comment is basically the offset (the characters between the last item/comment) to the current comment
+fn get_comments_between_spans(
+    from: &CommentSpan,
+    to: &CommentSpan,
+    comment_map: &CommentMap,
+    unformatted_code: &Arc<str>,
+) -> Vec<CommentWithContext> {
+    let mut comments_with_context = Vec::new();
+    for (index, comment_tuple) in comment_map
+        .range((Included(from), Excluded(to)))
+        .enumerate()
+    {
+        if comments_with_context.is_empty() {
+            // This is the first comment in the current range the context should be collected between from's end and comment's beginning
+            comments_with_context.push((
+                comment_tuple.1.clone(),
+                unformatted_code[from.end..comment_tuple.0.start].to_string(),
+            ));
+            println!("this");
+        } else {
+            // There is a comment before this one, so we should get the context starting from the last comment's end to the beginning of the current comment
+            comments_with_context.push((
+                comment_tuple.1.clone(),
+                unformatted_code
+                    [comments_with_context[index - 1].0.span.end()..comment_tuple.0.start]
+                    .to_string(),
+            ));
+        }
+    }
+    println!(
+        " found {:?} between {:?} and {:?} \n\n",
+        comments_with_context, from, to
+    );
+    comments_with_context
+}
+
+/// Inserts after given span and returns the offset. While inserting comments this also inserts Context of the comments so that the alignment whitespaces/newlines are intact
+fn insert_after_span(
+    from: &CommentSpan,
+    comments_to_insert: Vec<CommentWithContext>,
+    offset: usize,
+    formatted_code: &mut FormattedCode,
+) -> Result<usize, FormatterError> {
+    let mut src_rope = Rope::from_str(formatted_code);
+    // prepare the comment str
+    let mut comment_str = format!(
+        "{}{}",
+        comments_to_insert[0].1,
+        format_comment(&comments_to_insert[0].0)
+    );
+    for comment in comments_to_insert.iter().skip(1) {
+        write!(comment_str, "{}{}", comment.1, &format_comment(&comment.0))?;
+    }
+    src_rope.insert(from.end + offset, &comment_str);
+    formatted_code.clear();
+    formatted_code.push_str(&src_rope.to_string());
+    Ok(comment_str.len())
+}
+
+/// Applies formatting to the comment.
+/// Currently just checks if it is a multiline comment, if that is the case it adds a trailing `/` to the end.
+fn format_comment(comment: &Comment) -> String {
+    if comment.span().str().starts_with("/*") {
+        format!("{}/", comment.span().str())
```

Comment:
Why do we add the trailing '/' here? Is this missing from from the original comment span? If so, we should likely fix the span in the `lex_comments` function.

Reply:
Yes, it is missing in the original comment span. I didn't bring it up because it had a test without the trailing `/` so I thought it was intentional 😄  I just created #2355 for tracking this issue. Would you prefer that handled separately or should we pass it with this one?

Reply:
Agh I just took a look and that `assert!` is definitely wrong! I must have just copy pasted it from the output without thinking :facepalm:

I think we should address it before landing more code that works around it - maybe in a separate independent PR?